File: src/main/java/io/alpere/crudfop/service/ArchiveServiceImpl.java
package io.alpere.crudfop.service;

import io.alpere.crudfop.exception.EntityNotExistsException;
import io.alpere.crudfop.model.BaseEntity;
import io.alpere.crudfop.repository.BaseRepository;

import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import com.querydsl.core.types.dsl.BooleanExpression;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Getter
public class ArchiveServiceImpl<ArchivedEntity extends BaseEntity> implements ArchiveService<ArchivedEntity> {
    protected final BaseRepository<ArchivedEntity> repository;
    protected final Class<?> entityClass;

    public ArchiveServiceImpl(BaseRepository<ArchivedEntity> repository, Class<?> entityClass) throws Exception {
        if (!isBaseEntity(entityClass)) {
            throw new Exception("Error service initialization. Class not extend BaseEntity.");
        }
        this.repository = repository;
        this.entityClass = entityClass;
    }

    @Override
    public ArchivedEntity findOne(UUID id) {
        notNull(id);
        return repository.findById(id)
                .orElseThrow(() -> new EntityNotExistsException(entityClass.getSimpleName(), id));
    }

    @Override
    public List<ArchivedEntity> findAll() {
        return repository.findAll();
    }

    @Override
    public List<ArchivedEntity> findAll(BooleanExpression predicate) {
        if (predicate != null) {
            return newArrayList(repository.findAll(predicate));
        }
        return findAll();
    }

    @Override
    public ArchivedEntity restore(ArchivedEntity archivedEntity) {
        notNull(archivedEntity);
        if (exists(archivedEntity.getId())) {
            archivedEntity.setDeletedAt(null);
            archivedEntity.setDeletedBy(null);
            repository.save(archivedEntity);
        }
        return archivedEntity;
    }

    @Override
    public ArchivedEntity restore(UUID id) {
        return restore(findOne(id));
    }

    @Override
    public void restore(Iterable<ArchivedEntity> entities) {
        entities.forEach(this::restore);
    }

    @Override
    public void restoreAll(Iterable<UUID> ids) {
        notNullIds(ids);
        ids.forEach(this::restore);
    }

    @Override
    public void delete(ArchivedEntity archivedEntity) {
        notNull(archivedEntity);
        repository.delete(archivedEntity);
    }

    @Override
    public void delete(UUID id) {
        notNull(id);
        repository.deleteById(id);
    }

    @Override
    public void delete(Iterable<ArchivedEntity> archivedEntities) {
        notNull(archivedEntities);
        repository.deleteAll(archivedEntities);
    }

    @Override
    public void deleteAll(Iterable<UUID> ids) {
        notNullIds(ids);
        ids.forEach(this::delete);
    }

    public boolean exists(UUID id) {
        notNull(id);
        return repository.findById(id).isPresent();
    }

    private boolean isBaseEntity(Class entityClass) {
        return entityClass != BaseEntity.class && BaseEntity.class.isAssignableFrom(entityClass);
    }

    private void notNull(UUID id) {
        String message = String.format("Id for archived [%s] should not be null.", entityClass.getSimpleName());
        Objects.requireNonNull(id, message);
    }

    private void notNullIds(Iterable<UUID> ids) {
        String message = String.format("Ids for archived [%s] should not be null.", entityClass.getSimpleName());
        Objects.requireNonNull(ids, message);
    }

    private void notNull(ArchivedEntity archivedEntity) {
        String message = String.format("Archived entity [%s] should not be null.", entityClass.getSimpleName());
        Objects.requireNonNull(archivedEntity, message);
    }

    private void notNull(Iterable<ArchivedEntity> entities) {
        String message = String.format("List of [%s] should not be null.", entityClass.getSimpleName());
        Objects.requireNonNull(entities, message);
        entities.forEach(this::notNull);
    }

    private List<ArchivedEntity> newArrayList(Iterable<ArchivedEntity> elements) {
        notNull(elements);
        return StreamSupport.stream(elements.spliterator(), false)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/io/alpere/crud/service/ArchiveServiceImpl.kt
package io.alpere.crud.service

import com.querydsl.core.types.dsl.BooleanExpression
import io.alpere.crud.exception.EntityNotExistsException
import io.alpere.crud.model.BaseAuditEntity
import io.alpere.crud.repository.BaseRepository
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.*

open class ArchiveServiceImpl<ArchivedEntity : BaseAuditEntity>(
    protected open val repository: BaseRepository<ArchivedEntity>
) : AbstractService<ArchivedEntity>(), ArchiveService<ArchivedEntity> {

    @Transactional
    override fun findOne(id: UUID?): ArchivedEntity {
        return id.run(::safeId).let {
            repository.findById(it)
                .orElseThrow { EntityNotExistsException(entityName(), it) }
        }
    }

    @Transactional
    override fun findAll(): List<ArchivedEntity> {
        return repository.findAll()
    }

    @Transactional
    override fun findAll(predicate: BooleanExpression?): List<ArchivedEntity> {
        return predicate?.let {
            newArrayList(repository.findAll(it))
        } ?: findAll()
    }

    override fun restore(archivedEntity: ArchivedEntity?): ArchivedEntity {
        return archivedEntity.run(
            ::safeEntity
        ).takeIf {
            exists(it.id)
        }?.let {
            it.deletedAt = null
            it.deletedBy = null
            repository.save(it)
        } ?: throw EntityNotExistsException(entityName(), archivedEntity?.id)
    }

    override fun restore(id: UUID?): ArchivedEntity {
        return id.run(::findOne)
            .run(::restore)
    }

    override fun restore(archivedEntities: Iterable<ArchivedEntity?>?) {
        archivedEntities.run(::safeEntities).let { iterable ->
            iterable.forEach {
                restore(it)
            }
        }
    }

    override fun restoreAll(ids: Iterable<UUID?>?) {
        ids.run(::safeIds).let { iterable ->
            iterable.forEach {
                restore(it)
            }
        }
    }

    @Transactional
    override fun delete(archivedEntity: ArchivedEntity?) {
        archivedEntity.run(::safeEntity).let {
            repository.delete(it)
        }
    }

    @Transactional
    override fun delete(id: UUID?) {
        id.run(::safeId).let {
            repository.deleteById(it)
        }
    }


    @Transactional
    override fun delete(archivedEntities: Iterable<ArchivedEntity?>?) {
        archivedEntities.run(::safeEntities).let {
            repository.deleteAll(it)
        }
    }

    @Transactional
    override fun deleteAll(ids: Iterable<UUID?>?) {
        ids.run(::safeIds).let {
            repository.deleteAllById(it)
        }
    }

    override fun exists(id: UUID?): Boolean {
        return id.run(::safeId).let {
            repository.existsById(it)
        }
    }

    @Transactional
    override fun findFilteredByAudit(createdBy: String?, createdFrom: Instant?, createdTo: Instant?,
                                     updatedBy: String?, updatedFrom: Instant?, updatedTo: Instant?,
                                     deletedBy: String?, deletedFrom: Instant?, deletedTo: Instant?): List<ArchivedEntity> {

        return findAll(AuditFilter.createFilter(createdBy, createdFrom, createdTo,
            updatedBy, updatedFrom, updatedTo,
            deletedBy, deletedFrom, deletedTo))
    }

}